src/shared/tpm2-util.c

```diff
@@ -1158,23 +1164,30 @@ int tpm2_unseal(
                 goto finish;
         }
 
-        if (pin) {
-                TPM2B_AUTH auth = {
-                        .size = SHA256_DIGEST_SIZE
-                };
+        r = tpm2_make_encryption_session(c.esys_context, primary, hmac_key, pin, &hmac_session);
+        if (r < 0)
+                goto finish;
 
-                hash_pin(pin, strlen(pin), auth.buffer);
+        r = tpm2_make_pcr_session(
+                        c.esys_context,
+                        primary,
+                        hmac_session,
+                        TPM2_SE_POLICY,
+                        pcr_mask,
+                        pcr_bank,
+                        !!pin,
+                        &session,
+                        &policy_digest,
+                        /* ret_pcr_bank= */ NULL);
+        if (r < 0)
+                goto finish;
 
-                rc = sym_Esys_TR_SetAuth(c.esys_context, hmac_key, &auth);
-                explicit_bzero_safe(&auth, sizeof(auth));
-                if (rc != TSS2_RC_SUCCESS) {
-                        r = log_error_errno(
-                                        SYNTHETIC_ERRNO(ENOTRECOVERABLE),
-                                        "Failed to load PIN in TPM: %s",
-                                        sym_Tss2_RC_Decode(rc));
-                        goto finish;
-                }
-        }
+        /* If we know the policy hash to expect, and it doesn't match, we can shortcut things here, and not
+         * wait until the TPM2 tells us to go away. */
+        if (known_policy_hash_size > 0 &&
+            memcmp_nn(policy_digest->buffer, policy_digest->size, known_policy_hash, known_policy_hash_size) != 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EPERM),
+                                       "Current policy digest does not match stored policy digest, cancelling TPM2 authentication attempt.");
```

Comment:
@poettering I kept this hunk exactly the same, but it seems to suffer with indent issues as well, do you want me to correct that while I am here?

Reply:
well, i mean, sure, if you want ;-)